Update Quick Start Guide

hasheddan opened the pull request:
Signed-off-by: hasheddan <georgedanielmangum@gmail.com>

<!--
Thank you for helping to improve Crossplane!

We strongly recommend you look through our contributor guide at https://git.io/fj2m9
if this is your first time opening a Crossplane pull request. You can find us in
https://slack.crossplane.io/messages/dev if you need any help contributing.
-->

### Description of your changes
<!--
Briefly describe what this pull request does. Be sure to direct your reviewers'
attention to anything that needs special consideration.

We love pull requests that resolve an open Crossplane issue. If yours does, you
can uncomment the below line to indicate which issue your PR fixes, for example
"Fixes #500":

Fixes #
-->

WIP quick start for v0.11 docs

### How has this code been tested?
<!--
Before reviewers can be confident in the correctness of a pull request,
it needs to tested and shown to be correct. In this section, briefly
describe the testing that has already been done or which is planned.
-->

All examples need manual testing.

[skip ci]

### Checklist
<!--
Please run through the below readiness checklist. The first two items are
relevant to every Crossplane pull request.
-->
I have:
- [x] Run `make reviewable` to ensure this PR is ready for review.
- [x] Ensured this PR contains a neat, self documenting set of commits.
- [x] Updated any relevant [documentation] and [examples].
- [ ] Reported all new error conditions into the log or as an event, as
  appropriate.

For more about what we believe makes a pull request complete, see our
[definition of done].

[documentation]: https://github.com/crossplane/crossplane/tree/master/docs
[examples]: https://github.com/crossplane/crossplane/tree/master/cluster/examples
[definition of done]: https://github.com/crossplane/crossplane/tree/master/design/one-pager-definition-of-done.md

negz commented on docs/getting-started/provision-infrastructure.md:
```diff
@@ -0,0 +1,212 @@
+
+# Provision Infrastructure
+
+Crossplane's most basic feature is provisioning infrastructure anywhere using
```
Nit: Perhaps "core" or "foundation" rather than "most basic"?

negz commented:
@hasheddan Here's a Pod you can use to demonstrate that a PostgreSQL database is reachable:

```yaml
apiVersion: v1
kind: Pod
metadata:
  name: see-db
  namespace: default
spec:
  containers:
  - name: see-db
    image: postgres:9.6
    command: ['psql']
    args: ['-c', 'SELECT current_database();']
    env:
    - name: PGDATABASE
      value: postgres
    - name: PGHOST
      valueFrom:
        secretKeyRef:
          name: db-conn
          key: endpoint
    - name: PGUSER
      valueFrom:
        secretKeyRef:
          name: db-conn
          key: username
    - name: PGPASSWORD
      valueFrom:
        secretKeyRef:
          name: db-conn
          key: password
```

```console
$ kubectl apply -f pod.yaml 
pod/see-db created

$ kubectl logs see-db
 current_database 
------------------
 postgres
(1 row)
```

ryanzhang-oss commented on docs/getting-started/provision-infrastructure.md:
```diff
@@ -0,0 +1,224 @@
+
+# Provision Infrastructure
+
+Crossplane's core feature is provisioning infrastructure anywhere using the
+Kubernetes API. Once you have [installed a provider] and [configured your
```
I wonder how does this statement work with the adoption of an application model such as OAM in crossplane?

negz replied:
I suspect this might be meaning core in the foundational/fundamental/lowest level sense. We'll update the wording to ensure equal footing is given to application models.

hasheddan replied:
Definitely, I updated it to say "Crossplane allows you to provision infrastructure anywhere using the Kubernetes API." since that is what this section is about :+1:

ryanzhang-oss commented on docs/getting-started/provision-infrastructure.md:
```diff
@@ -0,0 +1,224 @@
+    dbInstanceStorageInGB: 20
+    securityIPList: "0.0.0.0/0"
+    masterUsername: "test123"
+  writeConnectionSecretsToRef:
```
where is this defined? I can't seem to see in the [Alibaba provider](https://github.com/crossplane/provider-alibaba/tree/master/apis) types.

hasheddan replied:
@ryanzhang-oss this was adapted from https://github.com/crossplane/provider-alibaba/blob/master/cluster/examples/database/resource_class.yaml

negz replied:
@ryanzhang-oss It comes from the embedded `ResourceSpec`, which is common to all Crossplane managed resources.

* https://github.com/crossplane/provider-alibaba/blob/b4b04d/apis/database/v1alpha1/rds_types.go#L94
* https://github.com/crossplane/crossplane-runtime/blob/290de/apis/core/v1alpha1/resource.go#L175

hasheddan commented on docs/getting-started/publish-infrastructure.md:
```diff
@@ -0,0 +1,560 @@
+[GCP]: #gcp
+[Azure]: #azure
+[Alibaba]: #alibaba
+[next section]: consume-infrastructure.md
```
Note: dependent on @negz naming of this doc

hasheddan commented on docs/getting-started/provision-infrastructure.md:
````diff
@@ -0,0 +1,224 @@
+You can view the progress of provisioning the database with the following
+command:
+```
+kubectl get -f aws-rds.yaml
````
@muvaf `get` will only show the printer columns right?

muvaf replied:
Oh sorry I thought it was YAML output, which is what I do usually.

negz replied:
Is it worth including the `kubectl describe` output anywhere in this guide? I like the succinctness of `kubectl get`, especially when you're just waiting for a thing, but `kubectl describe` (with status, events, etc) communicates much more of a "hey this actually got provisioned" experience in my opinion.

That all said, I could be convinced that `kubectl describe` output would be too overwhelming for this guide.

hasheddan replied:
Would you want to `kubectl describe` and show the output here?

hasheddan replied:
We could have some output that is collapsible so we could say something like "For more information run `kubectl describe` and you will see output like the following:" <collapsed-content>. That being said, I feel like we keep going back and forth between "all users will know how to use the Kubernetes API" and "some users may not know how to use the Kubernetes API". If we are saying the former then I think users will naturally `kubectl describe` if they want more information.

negz replied:
Yeah, I agree it's safe to assume folks will know about `kubectl describe`. My suggestion is not so much because I'm worried the reader won't know they _can_ use `kubectl describe`, but because I think `kubectl describe` makes a stronger impression because (at least to me) there's more "wow factor" in seeing the spec and status of the resource updated to reflect the state in the cloud. That all said, it is quite a lot of output for a quickstart so I'm happy as long as it's demonstrated elsewhere in the docs.

negz replied:
Let's stick with `kubectl get` for now - we can show `describe` in the more detailed guides.

hasheddan commented on docs/getting-started/publish-infrastructure.md:
````diff
@@ -0,0 +1,560 @@
+```
+
+Once the database is provisioned on your provider of choice, you should see
+`STATUS: Bound` in the output when you run:
````
Oh dear 🤦‍♂️ haha

hasheddan replied:
Side note: we should remove binding status from printer columns and maybe from resources entirely in the future

muvaf replied:
Future is here waiting your review in respective repos.

hasheddan commented on docs/getting-started/publish-infrastructure.md:
````diff
@@ -0,0 +1,560 @@
+Create a file named `pod.yaml` with the following contents:
+```yaml
+apiVersion: v1
+kind: Pod
````
Hmm yeah I think that would make sense, except for I’m not sure if this `Pod` will get into a long `CrashLoopBackoff` waiting for the DB.

muvaf replied:
If we create it after the DB is ready like composition example, why would it get into `CrashLoopBackoff`?

hasheddan replied:
Aren’t you saying applying it all as one set of YAML?

muvaf replied:
I was saying that since Azure requires at least 3 resources for public access, we can deploy those in one big YAML. Others would stay same. After all is done, we can instruct user to create this `Pod` after they see all resources they deployed are ready.

hasheddan replied:
> I was saying that since Azure requires at least 3 resources for public access, we can deploy those in one big YAML. 

@muvaf I think I may not be following here. The Azure `Composition` has all 3 of these resources.

muvaf replied:
I meant to add the `Pod` to the other document that is about provisioning infrastructure. The thing that would hold us back from that is Azure's extra resource requirement and we can create those together with MySQLServer, manually, as opposed to Composition.

negz replied:
> Should we create this Pod in infra provisioning step, too?

No, I don't think we should. Let's keep that section super simple; you apply YAML, you get infra in your cloud console. In reality I'm not sure we should be encouraging folks to statically provision something like a MySQLInstance and use it from a pod without ever creating a `Requirement`.

hasheddan commented on docs/getting-started/provision-infrastructure.md:
````diff
@@ -0,0 +1,224 @@
+kubectl get -f aws-rds.yaml
+```
+
+When provisioning is complete, you should see `STATUS: Unbound` in the output.
````
Yeah I agree that would be more appropriate I don’t think it is shown in the printer columns for these resources :/ ... I could use `STATE` and just update it for each resource to whatever the cloud provider gives back

muvaf replied:
I've started dropping PRs in all providers to include `READY` and `SYNCED` printer columns in all resources.

hasheddan replied:
@negz feel free to merge this at-will: https://github.com/crossplane/provider-alibaba/pull/11

hasheddan commented:
@muvaf that would be great! Thank you :)

negz commented on docs/getting-started/publish-infrastructure.md:
````diff
@@ -0,0 +1,560 @@
+Create the `ClusterRole`:
+```
+kubectl apply -f mysql-cr.yaml
+```
````
Could we store this `ClusterRole` as `quick-start-role.yaml` or something and `kubectl apply -f https://raw.url/quick-start-role.yaml`, without showing its contents here. Obviously this step needs to happen for things to work, but I don't think we add much value by demonstrating RBAC in motion.

hasheddan replied:
I personally feel like it’s kind of weird to see everything then just `kubectl apply` this one thing. That being said, it is not a core concept and I get that the idea is simplicity here. It seems like we are already assuming folks are familiar with k8s concepts though so maybe it’s not adding much complexity?

negz replied:
If I understand correctly the purpose of this guide is take folks on a tour of Crossplane concepts, getting them up and running as quickly as possible. This `ClusterRole` is needed for the Crossplane concepts we're introducing to work, but it's not a Crossplane concept, so to me it doesn't seem worth the extra lines of YAML and cognitive burden on the reader. We show the ClusterRole inline and discuss it in more detail in the longer composition guide.

Given this is a quick start I'd err toward including an obvious but succinct note like:

> Note that Crossplane must be granted an RBAC role allowing it to manage your newly defined and published infrastructure. Run `kubectl apply -f https://github.com/quick-start-role.yaml` to grant Crossplane access to the resources we'll cover in this guide.

hasheddan replied:
That sounds reasonable to me! Will update 👍

negz commented on docs/getting-started/publish-infrastructure.md:
```diff
@@ -0,0 +1,560 @@
+- `InfrastructurePublication`: makes an `InfrastructureDefinition` available at
+  the namespace scope
+- `ClusterRole`: allows Crossplane to manage this newly defined infrastructure
+  type
```
I think we can gloss over (i.e. omit) this one here, since it's not a Crossplane concept.

hasheddan replied:
Will address this depending on final resolution of https://github.com/crossplane/crossplane/pull/1514#discussion_r426226676

negz commented on docs/getting-started/publish-infrastructure.md:
```diff
@@ -0,0 +1,560 @@
+Now that we have defined our new type of infrastructure (`MySQLInstance`) and
+created at least one composition that can satisfy it, we can create a
+`MySQLInstanceRequirement` in the namespace of our choosing. In each
+`Composition` we defined we added a `provider: <name-of-provider` label. In the
```
```suggestion
`Composition` we defined we added a `provider: <name-of-provider>` label. In the
```

hasheddan commented on docs/getting-started/provision-infrastructure.md:
```diff
@@ -0,0 +1,224 @@
+Crossplane's core feature is provisioning infrastructure anywhere using the
+Kubernetes API. Once you have [installed a provider] and [configured your
+credentials], you can create any infrastructure resource currently supported by
+the provider. Let's start by provisioning a database on your provider of choice.
```
Potentially yes.. I am not sure each of the providers can be boiled down to a few lines to configure though

hasheddan commented on docs/getting-started/provision-infrastructure.md:
````diff
@@ -0,0 +1,224 @@
+
+Create the `RDSInstance`:
+```
+kubectl apply -f aws-rds.yaml
````
SGTM

hasheddan commented:
@negz @muvaf I have addressed most of the comments here either with changes or with a response. Since it can be somewhat difficult to see what the tabs will look like, here are some screenshots:
![tabs-1](https://user-images.githubusercontent.com/31777345/82151903-90de8180-9823-11ea-961f-115fd6b70604.png)
![tabs-2](https://user-images.githubusercontent.com/31777345/82151904-91771800-9823-11ea-844f-3ca365fa6486.png)

These are dependent on https://github.com/crossplane/crossplane.github.io/pull/43

prasek commented on docs/getting-started/provision-infrastructure.md:
```diff
@@ -0,0 +1,194 @@
+kind: CloudSQLInstance
+metadata:
+  name: cloudsqlmysql
+specTemplate:
```
```suggestion
spec:
```

prasek commented on docs/getting-started/publish-infrastructure.md:
```diff
@@ -0,0 +1,506 @@
+`MySQLInstanceRequirement` in the namespace of our choosing. In each
+`Composition` we defined we added a `provider: <name-of-provider>` label. In the
+`MySQLInstanceRequirement` below we use the `compositionSelector` with
+`provider: aws` to match our AWS `Composition`. This can be adjusted to select
```
Suggest tabs for each provider, since I just grabbed the content and ran it without reading the instructions, or maybe have the `matchLabels.provider`: REPLACE_ME in the example below.

prasek commented on docs/getting-started/publish-infrastructure.md:
```diff
@@ -0,0 +1,506 @@
+    - base:
+        apiVersion: database.gcp.crossplane.io/v1beta1
+        kind: CloudSQLInstance
+        specTemplate:
```
```suggestion
        spec:
```

prasek commented on docs/getting-started/publish-infrastructure.md:
```diff
@@ -0,0 +1,506 @@
+              tier: db-n1-standard-1
+              dataDiskType: PD_SSD
+              dataDiskSizeGb: 10
+          writeConnectionSecretsToRef:
```
```suggestion
          writeConnectionSecretToRef:
```

prasek commented on docs/getting-started/publish-infrastructure.md:
```diff
@@ -0,0 +1,506 @@
+            engine: mysql
+            skipFinalSnapshotBeforeDeletion: true
+            publiclyAccessible: true
+          writeConnectionSecretsToRef:
```
```suggestion
          writeConnectionSecretToRef:
```

haven't tested AWS yet, but docs say so this:
https://doc.crds.dev/github.com/crossplane/provider-aws/config/crd/database.aws.crossplane.io_rdsinstances.yaml#spec-writeConnectionSecretToRef

prasek commented on docs/getting-started/publish-infrastructure.md:
```diff
@@ -0,0 +1,506 @@
+              family: Gen5
+            storageProfile:
+              storageMB: 20480
+          writeConnectionSecretsToRef:
```
```suggestion
          writeConnectionSecretToRef:
```

prasek commented on docs/getting-started/publish-infrastructure.md:
```diff
@@ -0,0 +1,506 @@
+            dbInstanceStorageInGB: 20
+            securityIPList: "0.0.0.0/0"
+            masterUsername: "myuser"
+          writeConnectionSecretsToRef:
```
```suggestion
          writeConnectionSecretToRef:
```

hasheddan commented:
@prasek Thanks for your review! Just pushed update to address each of your comments. Here is screenshot of tabs for `MySQLInstanceRequirement`:

![tab-4](https://user-images.githubusercontent.com/31777345/82160317-082f0800-985a-11ea-9c03-bbd2ae47d37c.png)

muvaf commented on docs/getting-started/publish-infrastructure.md:
````diff
@@ -0,0 +1,567 @@
+should see the following output after creating it if you run `kubectl logs
+see-db`:
+```
+ current_database 
````
I don't see any output for a while when I follow AWS. Then the following after 2-3 min:
```
psql: could not connect to server: Connection refused
	Is the server running on host "default-my-db-zdcs5-g6bvd.csguarq5weag.us-east-1.rds.amazonaws.com" (54.160.126.136) and accepting
	TCP/IP connections on port 5432?
```
Could it be related to database being `MySQL` and client container image postgres? We don't mount a port and MySQL's default is 3306.

muvaf replied:
I see the same behavior in Azure, too.

hasheddan replied:
Ah yes at some point I switched from Postgres to MySQL, will update to use a MySQL client

negz replied:
See https://github.com/crossplane/crossplane/pull/1514/files#r426487230 for a MySQL compatible example.

muvaf commented on docs/getting-started/publish-infrastructure.md:
```diff
@@ -0,0 +1,567 @@
+            storageProfile:
+              storageMB: 20480
+          writeConnectionSecretToRef:
+            name: crossplane-system
```
```suggestion
            namespace: crossplane-system
```

muvaf commented on docs/getting-started/publish-infrastructure.md:
```diff
@@ -0,0 +1,567 @@
+              endIpAddress: 0.0.0.0
+          reclaimPolicy: Delete
+          providerRef:
+            name: example
```
```suggestion
            name: azure-provider
```

muvaf commented on docs/getting-started/publish-infrastructure.md:
```diff
@@ -0,0 +1,567 @@
+          reclaimPolicy: Delete
+          providerRef:
+            name: example
+      patches:
+        - fromFieldPath: "metadata.uid"
+          toFieldPath: "spec.writeConnectionSecretToRef.name"
+          transforms:
+            - type: string
+              string:
+                fmt: "%s-mysql"
+        - fromFieldPath: "spec.storageGB"
+          toFieldPath: "spec.forProvider.storageProfile.storageMB"
+          transforms:
+            - type: math
+              math:
+                multiply: 1024
+      connectionDetails:
+        - fromConnectionSecretKey: username
+        - fromConnectionSecretKey: password
+        - fromConnectionSecretKey: endpoint
```
As far as I can see, you don't need `patches` or `connectionDetails`. In fact, it causes error to have `spec.writeConnectionSecretToRef.name` without namespace.

hasheddan replied:
It was a typo to have `name` instead of `namespace` in the base, I am not sure why we would need no patches here?

negz replied:
I think there is a copy pasta bug here - all of the patches and connection details here apply to the MySQLServerFirewallRule, but it looks like they were intended to apply to the MySQLServer.

hasheddan replied:
Ah my apologies! The code snippet with this comment did not show whether it was with the `MySQLServerFirewallRule` or the `MySQLServer`, but I now see that the patch and connection details were duplicated :+1:

muvaf commented on docs/getting-started/publish-infrastructure.md:
```diff
@@ -0,0 +1,567 @@
+    - base:
+        apiVersion: azure.crossplane.io/v1alpha3
+        kind: ResourceGroup
+        spec:
```
`spec.location` is a required field of `ResourceGroup`. We don't seem to have this validation in CRD.
```
'create failed: cannot create ResourceGroup: resources.GroupsClient#CreateOrUpdate:
      Invalid input: autorest/validation: validation failed: parameter=parameters.Location
      constraint=Null value=(*string)(nil) details: value can not be null; required
      parameter'
```

negz commented on docs/getting-started/publish-infrastructure.md:
```diff
@@ -0,0 +1,567 @@
+              capacity: 2
+              family: Gen5
+            storageProfile:
+              storageMB: 20480
```
Typically when a field like `storageMB` is required by the definition I prefer to omit it from the base, since it will always be overwritten by the value of the composite/requirement.

negz commented on docs/getting-started/publish-infrastructure.md:
```diff
@@ -0,0 +1,567 @@
+            - type: string
+              string:
+                fmt: "%s-mysql"
+        - fromFieldPath: "spec.storageGB"
```
```suggestion
        - fromFieldPath: "spec.parameters.storageGB"
```

negz commented on docs/getting-started/publish-infrastructure.md:
```diff
@@ -0,0 +1,567 @@
+metadata:
+  name: mysqlinstances.aws.database.example.org
+  labels:
+    provider: aws
```
Could you label all of these Compositions with something consistent like `guide: quickstart`? That way the OAM section that follows this one will be able to select on that consistent label and get a valid composition for whatever cloud guide the reader followed.

hasheddan replied:
sure thing!

negz commented on docs/getting-started/publish-infrastructure.md:
```diff
@@ -0,0 +1,567 @@
+              matchControllerRef: true
+            properties:
+              startIpAddress: 0.0.0.0
+              endIpAddress: 0.0.0.0
```
```suggestion
              endIpAddress: 255.255.255.254
```

negz commented on docs/getting-started/publish-infrastructure.md:
````diff
@@ -0,0 +1,567 @@
+Kubernetes is the `Pod`. Let's define a `Pod` that will show that we are able to
+connect to our newly provisioned database.
+
+```yaml
+apiVersion: v1
+kind: Pod
+metadata:
+  name: see-db
+  namespace: default
+spec:
+  containers:
+  - name: see-db
+    image: postgres:9.6
+    command: ['psql']
+    args: ['-c', 'SELECT current_database();']
+    env:
+    - name: PGDATABASE
+      value: postgres
+    - name: PGHOST
+      valueFrom:
+        secretKeyRef:
+          name: db-conn
+          key: endpoint
+    - name: PGUSER
+      valueFrom:
+        secretKeyRef:
+          name: db-conn
+          key: username
+    - name: PGPASSWORD
+      valueFrom:
+        secretKeyRef:
+          name: db-conn
+          key: password
````
```suggestion
```yaml
apiVersion: v1
kind: Pod
metadata:
  name: see-db
  namespace: default
spec:
  containers:
  - name: see-db
    image: mysql:5.7
    command: ['mysql']
    args: ['-u', '$(MYSQL_USER)', '-e', 'SELECT CURRENT_USER();']
    env:
    - name: MYSQL_HOST
      valueFrom:
        secretKeyRef:
          name: db-conn
          key: endpoint
    - name: MYSQL_USER
      valueFrom:
        secretKeyRef:
          name: db-conn
          key: username
    - name: MYSQL_PWD
      valueFrom:
        secretKeyRef:
          name: db-conn
          key: password
```

This should do the trick:

```
$ kubectl logs -f see-db
CURRENT_USER()
myadmin@%
```

hasheddan replied:
Thanks @negz. Apologies for getting this mixed up the first time, and thanks for providing another solution :)

hasheddan commented on docs/getting-started/provision-infrastructure.md:
```diff
@@ -0,0 +1,194 @@
+</div>
+<div class="tab-pane fade" id="azure-tab-1" markdown="1">
+The Azure provider supports provisioning an [Azure Database for MySQL] instance
+with the `MySQLServer` CRD it installs into your cluster.
```
Good idea :)

hasheddan commented on docs/getting-started/publish-infrastructure.md:
```diff
@@ -0,0 +1,567 @@
+
+After creating the `MySQLInstanceRequirement` Crossplane will provision a
+database instance on your provider of choice. Once provisioning is complete, you
+should see `READY: True` in the output when you run:
```
Added a note with some helpful commands to view various groups of resources :+1:

prasek commented:
@hasheddan this looks awesome in the rendered docs, but is there a way for this to render more cleanly on GitHub when viewing the docs pages there?

![image](https://user-images.githubusercontent.com/284840/82229455-4de1e400-98df-11ea-840a-fa8643abc85f.png)

hasheddan commented:
@prasek I think this was solved in most recent update when I added additional linebreaks. This is how it appears for me: 
![tabs-nice](https://user-images.githubusercontent.com/31777345/82230020-cb622000-98f0-11ea-8ab4-8c74a4ee5a40.png)

Could you try refreshing and let me know if it looks better for you?

prasek commented:
@hasheddan looks great, thanks!

hasheddan commented:
Just ran through `provider-alibaba` (with this fix: https://github.com/crossplane/provider-alibaba/pull/13) and made some updates. Everything working as expected for alibaba now :+1:

kasey commented on docs/getting-started/provision-infrastructure.md:
```diff
@@ -0,0 +1,218 @@
+
+<!-- Named Links -->
+
+[installed a provider]: install.md
```
Should this link to the quickstart version instead? ( docs/getting-started/install-configure.md )
